File: 2024/24.rb

map = {}
$<.read.each_line{
    m = _1.scan(/\w+/)
    map[m[0]] = [m[1].to_i] if m.length == 2
    map[m[3]] = m[..2] if m.length == 4
}

#sw = "z07,bjm,hsw,z13,skf,z18"
#sw.split(",").each_cons(2) { |a,b| map[a],map[b] = map[b],map[a] }

op = {AND: :&, OR: :|, XOR: :^}
xs, ys, zs = "xyz".chars.map { |c|
    map.keys.select { _1.start_with?(c) }.sort
}

ops = ""
calc = ->(v, r=0) {
    expr = map[v]
    isv = expr.length == 1

    sop = isv ? v : "(#{expr.join(' ')})"
    ops = r==0 ? v + ' = ' + sop : ops.sub(v, sop) if r < 3

    next expr[0] if isv
    next 0 if r > 50
    a, b, c = calc[expr[0], r+1], expr[1].to_sym, calc[expr[2], r+1]
    a.send(op[b], c)
}

run = ->(n = 50) {
    zs.sort.map.with_index.sum { |v, i|
        i <= n ? calc[v] * 2**i : 0
    }
}

puts "Part 1: #{run[]}"

swaps = []
(xs+ys).each{map[_1] = [0]}

xs[2..].each { |xkey|
    i = xkey[1..].to_i
    map["x%02d" % -~i] = [0]
    map[xkey] = [1]
    run[i]
    cnt = op.keys.map { ops.scan(/(?= #{_1} )/).size }
    next if cnt == [2, 1, 2]

    l = [xkey.tr(?x, ?z)]
    2.times { l += l.flat_map{map[_1]} }
    l.filter! { map[_1] && _1[0]!='x' && _1[0]!='y' }
    swaps += [l.uniq]
}

xs.each{map[_1] = [0]}

valid = []
check = ->() {
    p ["te", valid]
    xs.all? {|xkey|
        map[xkey] = [1]

        r = run[]
        n = 1 << xkey[1..].to_i
        p ["cf", valid, xkey, r,  n] if r != n
        p xs.map{map[_1]}.join if r != n
        p ys.map{map[_1]}.join if r != n
        p (ys.map{"0"}.join + r.to_s(2))[-ys.size..].reverse! if r != n

        map[xkey] = [0]
        r == n
    }
}

swap_res = []
p swap_combos = (swaps.size/2).times.map { [swaps[_1*2], swaps[_1*2+1]] }
findvalid = ->(i = 0) {
    break (swap_res << valid.dup if check[]) unless swap_combos[i]
    la, lb = swap_combos[i]
    next if la[0] == lb[0]

    e = la[0]
    va = e[1..].to_i
    xkey = "x" + e[1..]
    map[xkey] = [1]

    la.each { |a| lb.each { |b|
        next if b[0] == 'z'

        map[a], map[b] = map[b], map[a]
        begin
            next if (1 << va) != run[va+1]

            valid << [a, b]
            map[xkey] = [0]
            findvalid[i + 1]
            map[xkey] = [1]
            valid.pop
        end
        map[a], map[b] = map[b], map[a]
    } }
    map[xkey] = [0]
}

findvalid[]

p swap_res

# z07,bjm, hsw,z13, skf,z18 wkr,nvr

File: 2024/24.js
let [state, conns] = app.ReadFile("test.txt").split("\n\n")
map = {}

state.split('\n').forEach(l => {
    l = l.split(': ')
    map[l[0]] = Number(l[1])
})

conns.split('\n').forEach(ll => {
    l = ll.match(/(\w+) (\w+) (\w+) -> (\w+)/)
    if (!l) return alert("oops: " + ll)
    map[l[4]] = l.slice(1, 4)
})
allvars = new Set(Object.keys(map).filter(v => v[0] != 'x' && v[0] != 'y'))

swap = (va, vb) => {
    var t = map[va]
    map[va] = map[vb]
    map[vb] = t
}

// [["z07", "njc", "kbk", "mkm", "tgj"], ["z08", "btr", "bjm", "tgj", "mkm"], 
// ["z13", "pmv", "rbk", "wrg", "rtn"], ["z14", "vhb", "bpf", "vjf", "hsw"], 
// ["z18"], ["z19", "scd", "rrq", "tgm", "skf"], 
// ["z26", "qkf", "wkr", "kvp", "qsm"], ["z27", "bpb", "crd", "frt", "nvr"]]

sw = ["z07,bjm", "hsw,z13", "skf,z18"]//,"wkr,nvr"]
app.SetClipboardText(sw.join(',').split(',').sort())
sw.map(s => swap(...s.split(",")))

app.WriteFile('input.txt', '')
var ss = Object.keys(map).filter(v => v[0] == 'z').sort().map(v => [v])
ss.map(l => {
    do {
        rr = false
        l0 = []
        l1 = (l[0].replace(/\b(z\d+|[a-z]{3})\b/g, v => (rr = true, l0.push('\t' + v + ' => ' + map[v].join(' ')), '(' + map[v].join(' ') + ')')))
        l.unshift(l0.join('\n'))
        l.unshift(l1)
    } while (rr)
    app.WriteFile('input.txt', l.reverse().join('\n').replace(/AND/g, '&').replace(/XOR/g, '^').replace(/OR/g, '|') + '\n\n', "append")
})

op = { AND: (a, b) => a & b, OR: (a, b) => a | b, XOR: (a, b) => a ^ b }
ops = ""
oplist = []
varlist = []
okset = new Set()
function calc(v, b) {
    var expr = map[v]
    if (expr.length) varlist.push(v)
    var sop = expr.length ? '(' + expr.join(' ') + ')' : v
    if (b !== undefined) ops = v + '=' + sop
    else ops = ops.replace(v, sop)

    if (typeof expr == "number") return expr
    var ret = op[expr[1]](calc(expr[0]), calc(expr[2]))
    if (b !== undefined) oplist.push('ops: ' + ops)
    return ret
}

function run(n = 50) {
    oplist = []
    varlist = []
    var r = Object.keys(map).filter(l => l[0] == "z").sort().map((v, i) => i <= n ? calc(v, n) : 0)
    return r.reduce((a, b, i) => a + b * 2 ** i, 0)
}
fmt = (n, i = 2) => ('00' + n).slice(-i)

alert(run()) // part1

res = []
vars = Object.entries(map)
var nx = vars.filter(e => e[0][0] == 'x').length
var ny = vars.filter(e => e[0][0] == 'y').length

for (var i = 0; i < nx; i++) {
    map['x' + fmt(i)] = 0
    map['y' + fmt(i)] = 0
}

var lfaults = []
for (var i = 0; i < nx; i++) {
    map['x' + fmt(i)] = 1
    v = run(i)
    res.push([i, 'x', v, v == 2 ** i])
    map['x' + fmt(i)] = 0
    if (v == 1 << i) varlist.forEach(v => okset.add(v))
}
alert(res.join('\n'))
